### DISCUSSION TITLE
preserve retain variable between function calls

### DISCUSSION THREAD

**Discussion by ctw1000** [dotnet/csharplang#8038] [Language Ideas]
Is there (I'd like) an easy way to preserve a variable value between function calls ?

not a class variable declared outside the function.
And not wrapped in something like a mini class.

But something simple and non-verbose.

Like

class MyClass
{
    void AddFunction()
    {
        preserved int thisIntKeepsItsValue = 1;
        //
        thisIntKeepsItsValue++;
    }
}

Because I want the scope of it limited to the function.

My invented 'preserved' keyword here would give it the life-time of the class too,
ie it wouldn't be  'static'

**Comment by foxesknow** [comment]
What's wrong with just having a class level static variable?

**Comment by ctw1000** [comment]
Yes - sure - that would work.

but
I write in a mix of C / C++ / C# depending on the month.
In C or C++ we can happily put a static variable in a function.

C# disallows this but we can make a class variable instead.
(it can be 'static' but it doesn't even need to be so long as its
  got the lifetime of the class instance).
  
  
But isn't this a forum for saying we want more??!


In a large class I don't want the confusion of 
"which function uses this variable".
Where I've made a class variable ONLY because the language
has forced me to do it and its only relevant to this single
function that wants to retain a value between calls.
ie  I want the scope limited.

-

In C# I always end up sticking it below the function like this:

class MyClass
{
   void FirstFunction()
   {
   }
   
   void SecondFunction()
   {
       m_variableThatKeepsItsValue++;
   }
   //--- this stuff used in the above function only ---
       int m_variableThatKeepsItsValue = 1;
   //---
}


....  and to edit my own writings...    having a static is not the same as
giving it class lifetime so the static-var-inside-a-function that you can
have in C or C++ is still not as nice as it could be with a say
void MyFunc()
{
     preserved int  thisVarHasLifeTimeOfTheOwningClassInCplusplusOrCsharp___orIsBasicallyStaticIfItsIn_C = 1;
}

(what if you want multiple instances of the class,   each with one of these....)

So I want this feature in C++ as well !

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/discussions/832

**Reply by TahirAhmadov** [reply]
Specifically, there is discussion in that thread about method-local instance fields, which is what I think the OP is asking for.

**Comment by MillKaDe** [comment]
> ... And not wrapped in something like a mini class. ...

Try a mini struct instead: [code at SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGJSqgOgGEIA7XCAGxgG4BYAKAEA3bFHL4wVcgF5yLGAHdyAWQCejDnlzkAFAEo+/EWIkAmGXMUr1m3Nv2GJDAMoxILACbEArBgCWrLoGAk70ru5evgEsQY6SYW6skf6BDiFgpgkRPikxafxmWUk50bHCouRCThahAOIwGABqoobGlWY1GfT1TS0CAOpQfhgwADJ+8roAJABE1bgAFhAArhwe5MAw5ADMIOQA3lWSAL4zwfyDw2MTWzqzHYsraxtbpnuHZqfnAsTblOZqDRaA4CciUP7EFDkABifiguAwJVSegO5GOoPICKgyzAGHI4WKvmgIP4YLBEzxxgAKgtsBgANIwGAAB1wAEkMLhmhxljAAEINBRMliMbAcDi4Qxk8gUmBQFhiyhQgmeJF5FH7DHS6m0hlM1kcrli3kCjBCmAisUSgDU1qlZOO5C14JlLDxPWaYmkAD5KqIaXTGSz2ZzuSbBcLReLJRj0aT8YlVUSxLgAGbcMEY35KhPZKLImS+tNFJO5Mrx7MU8geio+zGp7oNT2GY5AA==)

```c#
using static System.Console;

var mc1 = new MyClass ();
var mc2 = new MyClass ();
mc1.SecondFunction ();
mc1.SecondFunction ();
mc1.SecondFunction ();
mc2.SecondFunction ();
mc2.SecondFunction ();
var vmc1 = mc1.GetVar;
var vmc2 = mc2.GetVar;
WriteLine ($"mc1 should be 3: {vmc1}");
WriteLine ($"mc2 should be 2: {vmc2}");

public class MyClass {
  public void FirstFunction () { }
  struct SecondFunctor {
    int varThatKeepsItsValueBetweenCalls;
    internal void SecondFunction () {
      varThatKeepsItsValueBetweenCalls++;
    } 
    public int GetVar => varThatKeepsItsValueBetweenCalls;
  }
  SecondFunctor sf;  
  public void SecondFunction () => sf.SecondFunction ();
  public int GetVar => sf.GetVar;
}
```

This wraps the 'SecondFunction' and its 'varThatKeepsItsValueBetweenCalls' in an embedded struct 'SecondFunctor'.
Since it is a struct, there will be no extra heap allocation.

If you check out the generated code (click SharpLab link above) in **release mode**, you will notice that the Jitter inlined the mini struct 'SecondFunctor', its variable 'varThatKeepsItsValueBetweenCalls' and the struct member function 'SecondFunction'.
The generated code is pretty much the same that would have been generated without that inner struct 'SecondFunctor', so there is no runtime penalty.

But in the source, you have full control about who has access to what.

And in C/C++ it should work too.

**Reply by ctw1000** [reply]
Yes you've solved the problem of a variable keeping its value.

But why is this any improvement on having a private integer class member?


The problem I want the answer to 
besides being a short  non-verbose solution, ie  'not wrapping' something up)

- is to prevent anything else from inside of the class accessing this integer.


In your example I can still change the variable
'varThatKeepsItsValueBetweenCalls'
from anywhere in the class through  'sf'.
-
as I've done now in 'FirstFunction'  below



public class MyClass
{
  public void FirstFunction ()
  {
    sf.SecondFunction();    //<----  #####  we can still access sf' from here.   #####
  }
  
  
  struct SecondFunctor 
  {
    int varThatKeepsItsValueBetweenCalls;
    internal void SecondFunction () {
      varThatKeepsItsValueBetweenCalls++;
    } 
    public int GetVar => varThatKeepsItsValueBetweenCalls;
  }
  
  SecondFunctor sf;  
  
  public void SecondFunction () => sf.SecondFunction ();
  
  public int GetVar => sf.GetVar;
}



Perhaps your solution makes it visually clear that I shouldn't be accessing
it from elsewhere as you have to access something called 'secondFunctor'.
(suggesting it is related to secondFunction)
- ok.
And also I can only do to it what the SecondFunctor() func can do to it - in this
case  add 1.       - ok

But it still doesn't STOP anyone from calling this func and effecting it.

**Reply by HaloFour** [reply]
If you want to create a new form of accessibility you can use an analyzer to do so.

**Comment by MillKaDe** [comment]
> ...  is to prevent anything else from inside of the class accessing this integer.
> In your example I can still change the variable
> 'varThatKeepsItsValueBetweenCalls'
> from anywhere in the class through 'sf'.

No, you cannot.
'varThatKeepsItsValueBetweenCalls' is a **private** field in (nested) struct 'SecondFunctor', so only member functions or properties of  struct 'SecondFunctor' are able to read or write that field.
See [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGJSqgOgGEIA7XCAGxgG4BYAKAEA3bFHL4wVcgF5yLGAHdyAWQCejDnlzkAFAEo+/EWIkAmGXMUr1m3Nv2GJDAMoxILACbEArBgCWrLoGAk70ru5evgEsQY6SYW6skf6BDiFgpgkRPikxafxmWUk50bHCouRCThahAOIwGABqoobGlWY1GfT1TS0CAOpQfhgwADJ+8roAJABE1bgAFhAArhwe5MAw5ADMIOQA3lWSAL4zwfyDw2MTWzqzHYsraxtbpnuHZqfnAsTblOZqDRaA4CciUP7EFDkABifiguAwJVSegO5GOoPICKgyzAGHI4WKvmgIP4YLBEzxxgAKgtsBgANIwGAAB1wAEkMLhmhxljAAEINBRMliMbAcDi4Qxk8gUmBQFhiyhQgmeJF5FH7DHS6m0hlM1kcrli3kCjBCmAisUSgDU1qlZOO5C14JlLDxPWaYmkAD5KqIaXTGSz2ZzuSbBcLReLJRj0aT8YlVUSxLgAGbcMEY35KhPZKLImS+tNFJO5Mrx7OQ8g0uHJUr6EnaioICzFnWB/Uho08/kRi1RiX2sEAemH5DlUGJjGcpCopje5AA5IDbLgS8loPR23rg4aw72zZGrbhFzLtBNsGAwDA7H5gFxyB5eeQMBAZZzyMzJyNcaUuEIYA4WMsz+ClyA9CofUxVNugaT1DGOIA==)

```c#
  public void ThirdFunction () {
    var x = sf.varThatKeepsItsValueBetweenCalls;
    // error CS0122: 'MyClass.SecondFunctor.varThatKeepsItsValueBetweenCalls' is inaccessible due to its protection level
  }
```

> public class MyClass { public void FirstFunction () { sf.SecondFunction(); //<---- ##### we can still access sf' from here. ##### }

That is an indirect access through struct member function 'SecondFunctor.SecondFunction ()', so it is a controlled access.
You can neither read nor write 'varThatKeepsItsValueBetweenCalls' directly.
Of course, you can trigger incrementing it as a side effect of calling 'MyClass.SecondFunction ()', **exactly as in your original code**.

> Perhaps your solution makes it visually clear that I shouldn't be accessing it from elsewhere as you have to access something called 'secondFunctor'. (suggesting it is related to secondFunction)
> But it still doesn't STOP anyone from getting at it.

No, its not only visual - the compiler blocks any access from outside of nested struct 'secondFunctor'.

**Reply by ctw1000** [reply]
You are not wrong.
But I hope you would otherwise admit that while there is a partial
solution to the problem it is not a nice as it could be from a programmer's
point of view.   ?
Hoping someone will take this idea and make a simple keyword for doing
it to save me lots of typing in years to come.    :|

**Reply by HaloFour** [reply]
You can add your input to the [linked discussion](https://github.com/dotnet/csharplang/discussions/832).  However, as that discussion notes, it's not as simple as allowing `static` variables.  There are multiple policy concerns around how such a feature would behave, both in terms of lifetime and threading.

**Reply by ctw1000** [reply]
(edited to remove the giant text that you are about to complain about...     sorry for that)


I am not suggesting to make it static!
Well,  personally I would prefer to have the language allow anything - so yes -  but we
could choose not to use the feature of function static vars.
.
I am instead proposing that say a  'preserved' / 'retained' / 'owner-owned' function-variable
that has the lifetime of the object that it belongs to.
Not static.
Therefore not being global, not having memory initialized at app start-up,  and eliminating
cross-thread access issues.
.
Imagine it as a private class variable that has the scope of exactly one function.

**Reply by TahirAhmadov** [reply]
@ctw1000 if you look at https://github.com/dotnet/csharplang/discussions/832, another part of the conversation there is "method-local instance fields" - exactly what you are looking for.

**Reply by ctw1000** [reply]
Wow that was a long read.
Yes it seemed to touch on this idea at one point but for the most part it discusses 'static' vars rather than per-class-instance.

and hilariously the conversation ends with the comment:

HaloFour on Dec 2, 2023
"I think this is just going in circles and I don't know that opinions are changing at all, so it's probably best to leave it as it is. As it stands, none of this is championed even after at least nine years, and the more complicated it gets the less likely that seems like it would be."

:)

**Reply by CyrusNajmabadi** [reply]
@ctw1000 could you refrain from the large bolded responses :)   It comes across like you're shouting :)  Thanks very much!

**Reply by ctw1000** [reply]
Um.
The large font was an accident in the above case.
I thought it weird when it appeared large but then maybe it was because
it does this with the most recent comment not because of my font.

But it seems I did it somehow.  I am not shouting!

(that is a calm and quiet exclamation mark too.  haha).

**Reply by ctw1000** [reply]
Ah!  Just experimented.   if you put a single dash character between the paragraphs it makes everything bold.
Markup etc.      Accident.

**Reply by CyrusNajmabadi** [reply]
Thanks very much for fixing it! :)

**Reply by TahirAhmadov** [reply]
@ctw1000 if you add your voice to https://github.com/dotnet/csharplang/discussions/832 asking for method-scope instance locals, it would raise the chances of that proposal becoming championed, albeit it wouldn't guarantee it obviously.
Also it would be great if you proposed answers to some of the detail questions, such as, when is the local initialized, what happens if the same method is called from multiple threads, etc.

**Reply by ctw1000** [reply]
OK - just done it.  :)
(at the bottom)
